Feat/flagship certification

Ldoppea opened the pull request:
Add ability to register to `cozy-stack` as a Cozy flagship app

Consumer app can now init `cozy-client` with
`oauth.shouldRequireFlagshipPermissions` set to `true` in order to
register itself as a flagship app

When doing this, the OAuth process can now query respective platform's
stores in order to get an attestation of genuineness that can be
analysed by `cozy-stack`

This mechanism requires to be in a ReactNative environment

____

Related PR : https://github.com/cozy/cozy-react-native/pull/56
____

Todo:

- [x] Implement Android SafetyNet
- [x] Implement iOS DeviceCheck
- [x] Document the process
- [x] Drop commit `tests: TEMPORARY LOGS`

nono commented on packages/cozy-client/src/flagship-certification/README.md:
```diff
@@ -0,0 +1,49 @@
+- `flagship`: refers to the Cozy main application
+- `store certification`: process of verifying app's genuineness through the platform's store 
+- `attestation`: result from the `store certification`
+- `challenge`: unique token given to the app by `cozy-stack` (or by the store in iOS) that may be encrypted in the `attestation` as a proof of authenticity
```
s/(or by the store in iOS)//

Ldoppea replied:
Fixed 👍

nono commented on packages/cozy-client/src/flagship-certification/README.md:
```diff
@@ -0,0 +1,49 @@
+
+## iOS certification
+
+iOS certification is based on [DeviceCheck](https://developer.apple.com/documentation/devicecheck).
```
I'm not sure if we should speak of DeviceCheck or of AppAttest.

Ldoppea replied:
I kept the link to https://developer.apple.com/documentation/devicecheck but I now only refer to `AppAttest` (`DeviceCheck` is not mentioned anymore)

nono commented on packages/cozy-client/src/flagship-certification/flagship-certification.js:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Request a challenge from the Stack that can be used to request the app attestation from the app store
+ * @param {any} stackClient - the CozyStackClient instance
+ * @returns {Promise<string>} - the Nonce string retruned by the stack
```
s/retruned/returned/

Ldoppea replied:
Fixed 👍

nono reviewed (commented):
Are we sure we want to put this code in cozy-client? It will only be used for the flagship app, and it adds some dependencies.

Ldoppea commented:
I force-pushed the following:
- Refactored `getAppAttestationFromStore` to return the current platform, the attestation and an option KeyID (from iOS)
  - This allow `giveAppAttestationToStack` to be independent from the platform, it just send to the stack what was returned by `getAppAttestationFromStore`
- Refactored `certifyFlagship` to take a `certificationConfig` argument
  - For now it is used only to pass the `androidSafetyNetApiKey` from the flagship app (it should not be set directly on cozy-client's code)
- Added `typedefs.js` file for JSDoc types declaration
- Added `Client's configuration` and `Stacks' configuration` sections on the `README.md` documentation

Ldoppea commented:
> Are we sure we want to put this code in cozy-client? It will only be used for the flagship app, and it adds some dependencies.

I'm not sure how to extract this code from cozy-client. My initial understanding was that I had to put the code after `stackClient.register()` but before the `authorize()` step. But since our last conversation I'm not sure anymore that doing it before `authorize()` is a requirement.

A solution would be to inject a function pointer that would to the work on the flagship app.

However here is my opinion:
- This is fully related to cozy-client's OAuth registration, so it seems natural for me to have all related method in cozy-client's
  - Putting the code on the flagship side would segment the OAuth process too much (but this is a subjective point of view)
- The stack implementation uses arrays for all package names, certificates and app ids
  - This makes it easy to configure additional `cozy flagship` apps
  - So it seems more homogeneous to do a "generic" implementation on `cozy-client` too
  - Even if there is a few chances that we will do another flagship later, this would ease the work for auto-heberged cozy that would want to benefits from this feature on their own mobile app.
- This is a peerDependency, so until we are in a web context there is no need to install it. It will be required only in ReactNative contexts

Ldoppea commented:
I completed the doc to describe how to fill `apk_certificate_digests` using Android's local dev certificate: https://github.com/cozy/cozy-client/pull/1094/commits/b235e3f257cac4c2eabbfe010037d231ad3dceee

Crash-- commented on packages/cozy-client/package.json:
```diff
@@ -66,7 +66,9 @@
-    "react-native-inappbrowser-reborn": "^3.5.1"
+    "react-native-google-safetynet": "0.0.4",
+    "react-native-inappbrowser-reborn": "^3.5.1",
+    "react-native-ios11-devicecheck": "0.0.3"
```
should be our fork, right?

Ldoppea replied:
[Fixed](https://github.com/cozy/cozy-client/compare/1100dc38c7a24690f485b99a53b421a7d3e53b09..876e2ca52277c285a8e2bed5437fa6fa6ee58cc7)
